mono/metadata/mono-perfcounters.c

```diff
@@ -1591,98 +1603,123 @@ mono_perfcounter_instance_exists (MonoString *instance, MonoString *category)
 }
 
 /* this is an icall */
-MonoArray*
-mono_perfcounter_category_names (void)
+MonoArrayHandle
+mono_perfcounter_category_names (MonoError *error)
 {
-	ERROR_DECL (error);
+	HANDLE_LOOP_PREPARE;
+
 	int i;
-	MonoArray *res;
+	MonoArrayHandle res;
 	MonoDomain *domain = mono_domain_get ();
 	GSList *custom_categories, *tmp;
+
 	perfctr_lock ();
+
 	custom_categories = get_custom_categories ();
-	res = mono_array_new_checked (domain, mono_get_string_class (), NUM_CATEGORIES + g_slist_length (custom_categories), error);
+	res = mono_array_new_handle (domain, mono_get_string_class (), NUM_CATEGORIES + g_slist_length (custom_categories), error);
+
 	if (!is_ok (error)) {
-		res = NULL;
+		res = NULL_HANDLE_ARRAY;
 		goto leave;
 	}
 
 	for (i = 0; i < NUM_CATEGORIES; ++i) {
 		const CategoryDesc *cdesc = &predef_categories [i];
-		MonoString *name = mono_string_new_checked (domain, cdesc->name, error);
+		MonoStringHandle name;
+		SETUP_ICALL_FRAME;
+		name = mono_string_new_utf8_len (domain, cdesc->name, cdesc->name_length, error);
+		goto_if_nok (error, clear_icall_frame1);
```

Comment:
The `goto` carefully chosen label is a little ugly.
Alternatives:
 `_assign` which has other optimizations but was rejected
put success path under `if success` and so error path falls to the label.